sdk/cosmos/azure-cosmos-spark_3-2_2-12/src/main/scala/com/azure/cosmos/spark/CosmosWriter.scala

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.spark
+
+import com.azure.cosmos.CosmosDiagnosticsContext
+import com.azure.cosmos.implementation.ImplementationBridgeHelpers
+import org.apache.spark.SparkInternalsBridge
+import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.connector.metric.CustomTaskMetric
+import org.apache.spark.sql.connector.write.WriterCommitMessage
+import org.apache.spark.sql.execution.metric.CustomMetrics
+import org.apache.spark.sql.types.StructType
+
+import java.util.concurrent.atomic.AtomicLong
+
+private class CosmosWriter(
+                            userConfig: Map[String, String],
+                            cosmosClientStateHandles: Broadcast[CosmosClientMetadataCachesSnapshots],
+                            diagnosticsConfig: DiagnosticsConfig,
+                            inputSchema: StructType,
+                            partitionId: Int,
+                            taskId: Long,
+                            epochId: Option[Long],
+                            sparkEnvironmentInfo: String)
+  extends CosmosWriterBase(
+    userConfig,
+    cosmosClientStateHandles,
+    diagnosticsConfig,
+    inputSchema,
+    partitionId,
+    taskId,
+    epochId,
+    sparkEnvironmentInfo
+  ) with OutputMetricsPublisherTrait {
+
+  private val recordsWritten = new AtomicLong(0)
+  private val bytesWritten = new AtomicLong(0)
+  private val totalRequestCharge = new AtomicLong()
```

Comment:
Yes, Spark 3.1 does not have a concept of custom write metrics - so, there is simply no way to produce this custom metric. In Spark 3.2 and above I added it for two reaosns - first, I think it is a useful metric. Second, I wanted to explore the path of custom metrics in case we want to add others eventuall.y